infrastructures/openshift/src/main/java/org/eclipse/che/workspace/infrastructure/openshift/provision/InstallerServersPortProvisioner.java

```diff
@@ -0,0 +1,178 @@
+import org.eclipse.che.commons.lang.Pair;
+import org.eclipse.che.workspace.infrastructure.openshift.Names;
+import org.eclipse.che.workspace.infrastructure.openshift.environment.OpenShiftEnvironment;
+
+/**
+ * Fixes installers servers ports if conflicts are present.
+ *
+ * <p>Installers fail to start if they provide servers which should be running on the same port.
+ * Since containers in a pod uses shared ports installers in different machine can conflict too. To
+ * resolve such conflicts Infrastructure provides free port for installer server by injection
+ * environment variable.
+ *
+ * <p>The environment variable name has the following format
+ * `CHE_SERVER_{NORMALISED_SERVER_NAME}_PORT`. Where `NORMALISED_SERVER_NAME` is a server name in
+ * upper case where all character which are not latin letter or digit replaced with `_` symbol. So
+ * installers must respect the corresponding environment variables with the recommended value for
+ * server port.
+ *
+ * @author Sergii Leshchenko
+ */
+public class InstallerServersPortProvisioner implements ConfigurationProvisioner {
+
+  public static final String SERVER_PORT_ENV_FMT = "CHE_SERVER_%s_PORT";
+
+  private final int minPort;
+  private final int maxPort;
+
+  @Inject
+  public InstallerServersPortProvisioner(
+      @Named("che.infra.openshift.installer_server_min_port") int minPort,
+      @Named("che.infra.openshift.installer_server_max_port") int maxPort) {
+    this.minPort = minPort;
+    this.maxPort = maxPort;
+  }
+
+  @Override
+  public void provision(OpenShiftEnvironment osEnv, RuntimeIdentity identity)
+      throws InfrastructureException {
+    for (Pod pod : osEnv.getPods().values()) {
+      List<Container> containers = pod.getSpec().getContainers();
+      Map<String, InternalMachineConfig> podMachines = new HashMap<>();
+
+      for (Container container : containers) {
+        String machineName = Names.machineName(pod, container);
+        InternalMachineConfig machineConfig = osEnv.getMachines().get(machineName);
+        podMachines.put(machineName, machineConfig);
+      }
+
+      // containers in a pod use the same ports
+      fixInstallersPortsConflicts(podMachines);
+    }
+  }
+
+  @VisibleForTesting
+  void fixInstallersPortsConflicts(Map<String, InternalMachineConfig> machines)
+      throws InfrastructureException {
+    Set<Integer> occupiedPorts = new HashSet<>();
+    AtomicInteger freePort = new AtomicInteger(minPort);
+
+    for (InternalMachineConfig machineConfig : machines.values()) {
+      for (InstallerImpl installer : machineConfig.getInstallers()) {
+        Map<Integer, Collection<String>> port2ServerRefs = getServersRefsGroupedByPorts(installer);
+
+        for (Entry<Integer, Collection<String>> portServersEntry : port2ServerRefs.entrySet()) {
+          Integer port = portServersEntry.getKey();
+
+          if (!occupiedPorts.add(port)) {
+            int newPort = findFreePort(freePort, occupiedPorts);
```

Comment:
NIce proposal. But it will make this class even more complicated.
Because we can't just to add all servers ports to occupied ports while installers servers are already there. I'll try to skip installers servers and add result list ports to occupied ports list.

Reply:
I created new internal class. Please take a look.